### PR TITLE
mob: Lookahead optimizer (k=5/10, alpha=0.5/0.8)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The yi fleet has discovered that lr=5e-4 is the hard stability ceiling for AdamW on the PR #99 base config — any change that perturbs the gradient landscape causes divergence. Meanwhile, PR #144 (edward) is testing lower beta2 as a stability unlocker. This PR tests a complementary approach: **Lookahead** (Zhang et al., 2019 — https://arxiv.org/abs/1907.08610).

Lookahead maintains two sets of weights:
- **fast weights** θ: updated by the inner optimizer (AdamW) every step
- **slow weights** φ: interpolated toward fast weights every k steps: `φ ← φ + α(θ - φ)`

After the slow update, fast weights are reset to φ. This has two beneficial effects:
1. **Variance reduction**: fast weight oscillation around the loss minimum is averaged out by the slow step
2. **Stability**: the slow weight interpolation acts as a regularizer on the learning trajectory — effectively reducing the LR sensitivity of the inner optimizer. Runs that diverge with AdamW alone at lr=5e-4 may remain stable with Lookahead wrapping because the slow step damps extreme updates.

Lookahead was shown to work well with AdamW (the RAdam+Lookahead="Ranger" optimizer) and is known to find flatter minima with better generalization. Reference: "Lookahead Optimizer: k steps forward, 1 step back" (Zhang et al., NeurIPS 2019).

**Parameter choices to test:**
- `k=5, alpha=0.5` (Zhang et al. default) — most studied, moderate smoothing
- `k=10, alpha=0.8` — more aggressive slow-weight interpolation, stronger variance reduction

## Instructions

**Step 1 — Implement Lookahead wrapper** (self-contained, ~40 lines, no external deps):

```python
class Lookahead(torch.optim.Optimizer):
    """
    Lookahead optimizer: k steps forward, 1 step back.
    Wraps any inner optimizer. Zhang et al., NeurIPS 2019.
    https://arxiv.org/abs/1907.08610
    """
    def __init__(self, optimizer, k=5, alpha=0.5):
        self.optimizer = optimizer
        self.k = k
        self.alpha = alpha
        self.param_groups = optimizer.param_groups
        self.state = defaultdict(dict)
        self._step_counter = 0
        # Initialize slow weights
        for group in self.param_groups:
            for p in group['params']:
                self.state[p]['slow_buffer'] = p.data.clone()

    def step(self, closure=None):
        loss = self.optimizer.step(closure)
        self._step_counter += 1
        if self._step_counter % self.k == 0:
            for group in self.param_groups:
                for p in group['params']:
                    if p.grad is None:
                        continue
                    slow = self.state[p]['slow_buffer']
                    slow.add_(p.data - slow, alpha=self.alpha)
                    p.data.copy_(slow)
        return loss

    def zero_grad(self, set_to_none=True):
        self.optimizer.zero_grad(set_to_none=set_to_none)

    def state_dict(self):
        fast_state = self.optimizer.state_dict()
        slow_state = {i: v['slow_buffer'] for i, (k, v) in enumerate(self.state.items()) if isinstance(k, torch.Tensor)}
        return {'fast': fast_state, 'slow': slow_state, 'k': self.k, 'alpha': self.alpha}

    def load_state_dict(self, state_dict):
        self.optimizer.load_state_dict(state_dict['fast'])
        self.k = state_dict['k']
        self.alpha = state_dict['alpha']
```

Add `from collections import defaultdict` at top of train.py if not already present.

**Step 2 — Wire into train.py:**

Add to `Config`:
```python
use_lookahead: bool = False
lookahead_k: int = 5
lookahead_alpha: float = 0.5
```

After AdamW construction, optionally wrap:
```python
if config.use_lookahead:
    optimizer = Lookahead(optimizer, k=config.lookahead_k, alpha=config.lookahead_alpha)
```

Log `train/lookahead_k` and `train/lookahead_alpha` in W&B init.

**Note on EMA:** The existing EMA uses `model.parameters()` directly — this remains unchanged. Lookahead slow weights and EMA are orthogonal averaging schemes operating at different timescales and can coexist without modification.

**3-arm sweep:**

| Arm | k | alpha | lr | Run name |
|---|---:|---:|---|---|
| A | 5 | 0.5 | 5e-4 | `mob-lah-k5-a05-lr5e4` |
| B | 10 | 0.8 | 5e-4 | `mob-lah-k10-a08-lr5e4` |
| C | 5 | 0.5 | 3e-4 | `mob-lah-k5-a05-lr3e4` |

Arm C (lr=3e-4) is the stability fallback — if A or B diverge, C confirms that Lookahead stabilizes the lr=3e-4 regime.

```bash
cd target/
python train.py \
  --use-lookahead \
  --lookahead-k <5|10> \
  --lookahead-alpha <0.5|0.8> \
  --lr <5e-4|3e-4> \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 \
  --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group mob-lookahead-optimizer-r10 \
  --wandb-run-name <arm-name>
```

**What to watch:** If gradient norm spikes past 5 before clipping at step ~9000 (the known instability window), Lookahead did not stabilize lr=5e-4. If val abupt at epoch 2 is below 12.4 (faster than baseline's 12.42 at epoch 2), Lookahead is accelerating early convergence.

## Baseline (PR #99, W&B run `3hljb0mg`)

| Metric | yi best (val) | yi best (test) | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.69** | **11.73** | — |
| `surface_pressure_rel_l2_pct` | 6.97 | 6.64 | 3.82 |
| `wall_shear_rel_l2_pct` | 11.69 | 11.48 | 7.29 |
| `volume_pressure_rel_l2_pct` | 7.85 | 14.42 | 6.08 |
| `wall_shear_x_rel_l2_pct` | 10.17 | 10.06 | 5.35 |
| `wall_shear_y_rel_l2_pct` | **13.73** | **13.53** | **3.65** |
| `wall_shear_z_rel_l2_pct` | **14.73** | **13.98** | **3.63** |

PR #99 val trajectory: epoch 1=16.47, epoch 2=12.42, epoch 3=10.98, best=10.69.

Reproduce baseline:
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #153 because ADVISOR: Closing as orphaned. No senpai-yi-mob pod exists on this track and 'mob' is not in the active student roster, so this PR will never be picked up by an agent. Lookahead optimizer remains an interesting direction and can be reassigned to an active student in a future round..
